library/spdm_transport_storage_lib/libspdm_storage.c

```diff
@@ -0,0 +1,792 @@
+ *
+ * This value is transport layer specific.
+ *
+ * @param sequence_number        The current sequence number used to encode or decode message.
+ * @param sequence_number_buffer  A buffer to hold the sequence number output used in the secured message.
+ *                             The size in byte of the output buffer shall be 8.
+ *
+ * @return size in byte of the sequence_number_buffer.
+ *        It shall be no greater than 8.
+ *        0 means no sequence number is required.
+ **/
+uint8_t libspdm_storage_get_sequence_number(uint64_t sequence_number,
+                                            uint8_t *sequence_number_buffer)
+{
+    libspdm_copy_mem(sequence_number_buffer, LIBSPDM_STORAGE_SEQUENCE_NUMBER_COUNT,
+                     &sequence_number, LIBSPDM_STORAGE_SEQUENCE_NUMBER_COUNT);
+    return LIBSPDM_STORAGE_SEQUENCE_NUMBER_COUNT;
+}
+
+/**
+ * Decode an Security Protocol Command message to a normal message or secured message.
+ *
+ * @param  session_id                  Indicates if it is a secured message protected via SPDM session.
+ *                                     If *session_id is NULL, it is a normal message.
+ *                                     If *session_id is NOT NULL, it is a secured message.
+ * @param  connection_id               Indicates the connection ID of the message.
+ * @param  transport_message_size      size in bytes of the transport message data buffer.
+ * @param  transport_message           A pointer to a source buffer to store the transport message.
+ * @param  message_size                size in bytes of the message data buffer.
+ * @param  message                     A pointer to a destination buffer to store the message.
+ *
+ * @retval RETURN_SUCCESS                      The message is encoded successfully.
+ * @retval LIBSPDM_STATUS_INVALID_MSG_SIZE     The message is NULL or the transport_message_size is zero.
+ * @retval LIBSPDM_STATUS_INVALID_MSG_FIELD    The message field is incorrect.
+ **/
+libspdm_return_t libspdm_storage_decode_message(uint32_t **session_id,
+                                                uint8_t *connection_id,
+                                                size_t transport_message_size,
+                                                void *transport_message,
+                                                size_t *message_size,
+                                                void **message)
+{
+    const storage_spdm_transport_header *storage_header;
+    uint32_t storage_length;
+    uint16_t security_protocol_specific;
+    uint8_t spsp0, spsp1, spdm_operation;
+
+    LIBSPDM_ASSERT(transport_message_size >= sizeof(storage_spdm_transport_header));
+    if (transport_message_size <= sizeof(storage_spdm_transport_header)) {
+        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
+    }
+
+    if (transport_message_size == 0) {
+        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
+    }
+
+    storage_header = transport_message;
```

Comment:
I am confused here. If you say "It is not part of the SPDM message.", then why it points to transport_message and fill data.

Reply:
Ah let me clarify, by "It is not part of the SPDM message" I meant for example, for a request, that header does not go through spdm data buffer in the storage api, but instead the storage transport layer. For reference: [this](https://github.com/westerndigitalcorporation/spdm-utils/pull/139/files#diff-90dda305f29d7ba03588448fa907041edf34fae37e66558e2465dd555b00272cR360) is how it would be used by an upper level application using `libspdm`